### ISSUE TITLE
systemd-shutdown hangs on containerd-shim when k3s-agent running

### ISSUE THREAD

**Issue by sourcedelica** [rancher/k3s#2400]
**Environmental Info:**
K3s Version: k3s version v1.18.6+k3s1 (6f56fa1d)

Node(s) CPU architecture, OS, and Version:  x86_64 Ubuntu 20.04.1
Linux nuc-linux3 5.4.0-48-generic #52-Ubuntu SMP Thu Sep 10 10:58:49 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

Cluster Configuration: 1 master 2 workers

**Describe the bug:**
When shutting down or rebooting the node, the shutdown hangs for approximately 90 seconds.  The console message is
```
systemd-shutdown: waiting for process: containerd-shim
```

When researching the problem I landed on this issue: https://github.com/drud/ddev/issues/2538 where they said when they uninstalled k3s the problem went away.  I disabled and stopped `k3s-agent.service` and rebooted and the problem did go away for me.

**Comment by sourcedelica** [comment]
Following https://github.com/containerd/containerd/issues/386#issuecomment-304837687 I changed the service configuration for `k3s.agent` and `k3s-agent.service` to `KillMode=Mixed` and that fixed the problem.   This is in the standard Docker configuration.

However, I also found https://github.com/rancher/k3s/issues/1965 where it looks like this behavior is as intended.  Is there a way to allow for upgrading k3s without disrupting workloads but at the same time not hang shutdowns/reboots for 90s?

**Comment by sourcedelica** [comment]
I was thinking one way to do it is to use `KillMode=mixed` or `KillMode=control-group` by default for `k3s{-agent}.service` and when doing an upgrade, add a drop-in in `/run/systemd/system/k3s{-agent}.service.d` that temporarily sets `KillMode=process` before stopping the service, then removes the drop-in after the upgrade.

**Comment by dontlaugh** [comment]
systemd has an explicit pre-shutdown hook, so perhaps you could invoke special logic with that. See:

```
/usr/lib/systemd/system/shutdown.target.wants
```

**Comment by unixfox** [comment]
Bump still relevant

**Comment by senpaiSubby** [comment]
bump same issue

**Comment by andrewchen5678** [comment]
notice the issue with raspberry pi with display on

**Comment by jraby** [comment]
I faced this issue yesterday and ended up with the following solution.

`/etc/systemd/system/cgroup-kill-on-shutdown@.service` :
```
[Unit]
Description=Kill cgroup procs on shutdown for %i
DefaultDependencies=false
Before=shutdown.target umount.target
[Service]
# Instanced units are not part of system.slice for some reason
# without this, the service isn't started at shutdown
Slice=system.slice
ExecStart=/bin/bash -c 'pids=$(cat /sys/fs/cgroup/unified/system.slice/%i/cgroup.procs); echo $pids | xargs -r kill;'                                                                                                                                                                      
ExecStart=/bin/sleep 5                                                                                                                                                                                                                                                                     
ExecStart=/bin/bash -c 'pids=$(cat /sys/fs/cgroup/unified/system.slice/%i/cgroup.procs); echo $pids | xargs -r kill -9;'
Type=oneshot
[Install]
WantedBy=shutdown.target
```


Enable the "service" for `k3s-agent.service`  (will also work for `k3s` on the master ):
```
sudo systemctl enable cgroup-kill-on-shutdown@k3s-agent.service.service

# or, on the master:  sudo systemctl enable cgroup-kill-on-shutdown@k3s.service.service
```

I've written a long winding explanation [here](https://raby.sh/debugging-90s-hangs-during-shutdown-on-ubuntu-2004.html) but in brief, what happens is that since `killmode=process` is used, all the container processes end up staying alive when k3s is brought down. Which is a good thing :tm: 

However, during shutdown, systemd will signal all remaining processes and wait for `DefaultTimeoutStopSec` for them to die.
This is **always** 90s during the last shutdown phase with systemd v245.  
It is a bug in systemd v245 shipped with ubuntu 20.04 and was [fixed in september 2020](https://github.com/systemd/systemd/commit/7d9eea2bd3d4f83668c7a78754d201b22)

What I used to do was to set  `DefaultTimeoutStopSec=5s` in `/etc/systemd/system.conf` and it worked fine, but on ubuntu 20.04 it doesn't.

Since there's little chance this fix will make it back into 20.04, the above "service" will perform round of SIGTERM, wait 5s, then proceed with SIGKILL to finish k3s's process cleanup during shutdown.
The sleep can be tweaked to suit your services need (something matching `terminationGracePeriod` perhaps)

Hope it helps.

**Comment by sourcedelica** [comment]
Awesome research!

**Comment by miraculixx** [comment]
@jraby your solution helped me to resolve the issue, however I ended up using the `k3s-killall.sh` according to the [k3s docs ](https://rancher.com/docs/k3s/latest/en/upgrades/killall/) 

```
# source https://github.com/k3s-io/k3s/issues/2400#issuecomment-1013798094
# $ sudo systemctl enable cgroup-kill-on-shutdown@k3s-agent.service.service
[Unit]
Description=Kill cgroup procs on shutdown for %i
DefaultDependencies=false
Before=shutdown.target umount.target
[Service]
# Instanced units are not part of system.slice for some reason
# without this, the service isn't started at shutdown
Slice=system.slice
ExecStart=/bin/bash -c "/usr/local/bin/k3s-killall.sh"
Type=oneshot
[Install]
WantedBy=shutdown.target
```

This is on 

Linux Mint 20.2 
5.4.0-91-generic

**Comment by horihel** [comment]
the same problem exists on rke2 (no surprise, given it's roots are in k3s)

**Comment by brandond** [comment]
Yes, this is by design. Stopping the K3s (or RKE2) service does not stop running containers. This is to allow for nondisruptive upgrades of the main K3s/RKE2 components by simply replacing the binary and restarting the service.

**Comment by horihel** [comment]
would you accept a feature request to add a systemd unit like https://github.com/k3s-io/k3s/issues/2400#issuecomment-1018472343 which only triggers on shutdown?
This would both allow the intended behaviour of k3s/rke2 (seamless updates/restarts) _and_ allow for a shutdown/reboot that's even quicker than RKE1.

here's my non-instanced version of that:
```systemd-unit
[Unit]
Description=Kill containerd-shims on shutdown
DefaultDependencies=false
Before=shutdown.target umount.target

[Service]
ExecStart=/bin/bash -c "/usr/local/bin/rke2-killall.sh"
Type=oneshot

[Install]
WantedBy=shutdown.target
```

**Comment by brandond** [comment]
That might be a good thing to add to the documentation, for folks that want it?

**Comment by ciacon** [comment]
Confirming this behaviour to be present with:

```
root@k3s:~# k3s --version
k3s version v1.23.3+k3s1 (5fb370e5)
go version go1.17.5
root@k3s:~# uname -a
Linux k3s 5.4.0-100-generic #113-Ubuntu SMP Thu Feb 3 18:43:29 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by brandond** [comment]
This is not version-specific behavior.  As described at https://github.com/k3s-io/k3s/issues/2400#issuecomment-1040795816 by design, pods are not stopped when the systemd service is stopped.

**Comment by hlacikd** [comment]
in releases prior to 1.23.7 it was enough to add KillMode=mixed to /etc/systemd/system/k3s.service , and when system shutdown executed, k3s killed containers and computer was turned off imediately 

For some reason unknown to me since 1.23.8 ---> up to current one 1.24.4 when doing so, it takes 90s again to shutdown system with k3s (... which is default systemctl timeout TimeoutStopUSec=1min 30s ... ), so KillMode mixed is ignored and k3s waits until timeout has passed to kill them ....

what has changed?


```
[Service]
Type=notify
EnvironmentFile=-/etc/default/%N
EnvironmentFile=-/etc/sysconfig/%N
EnvironmentFile=-/etc/systemd/system/k3s.service.env
KillMode=mixed
```

**Comment by brandond** [comment]
Probably something related to the containerd version change? I'm not sure, since changing the KillMode isn't something we test or support. I would recommend adding another unit that runs on shutdown, as described above.

**Comment by hlacikd** [comment]
> Probably something related to the containerd version change? I'm not sure, since changing the KillMode isn't something we test or support. I would recommend adding another unit that runs on shutdown, as described above.

Thanks I have implemented shutdown unit as described in @horihel .

My I vote for adding this to official documentation? I believe it could be pretty common scenario, since k3s is ideal for edge deployments, and usually edge deployments get much more shutdowns then restarts.

**Comment by MountComb** [comment]
Here is a k3s version of https://github.com/k3s-io/k3s/issues/2400#issuecomment-1041165341:

```
[Unit]
Description=Kill containerd-shims on shutdown
DefaultDependencies=false
Before=shutdown.target umount.target

[Service]
ExecStart=/usr/local/bin/k3s-killall.sh
Type=oneshot

[Install]
WantedBy=shutdown.target
```

Put the file to `/etc/systemd/system/shutdown-k3s.service` and then enable the service using 
```
systemctl enable shutdown-k3s.service
```

Also note that this service name `shutdown-k3s` shall not start with `k3s-`, otherwise the `k3s-killall.sh` script would try to stop it and cause problems.

**Comment by samip5** [comment]
> I faced this issue yesterday and ended up with the following solution.
> 
> `/etc/systemd/system/cgroup-kill-on-shutdown@.service` :
> 
> ```
> [Unit]
> Description=Kill cgroup procs on shutdown for %i
> DefaultDependencies=false
> Before=shutdown.target umount.target
> [Service]
> # Instanced units are not part of system.slice for some reason
> # without this, the service isn't started at shutdown
> Slice=system.slice
> ExecStart=/bin/bash -c 'pids=$(cat /sys/fs/cgroup/unified/system.slice/%i/cgroup.procs); echo $pids | xargs -r kill;'                                                                                                                                                                      
> ExecStart=/bin/sleep 5                                                                                                                                                                                                                                                                     
> ExecStart=/bin/bash -c 'pids=$(cat /sys/fs/cgroup/unified/system.slice/%i/cgroup.procs); echo $pids | xargs -r kill -9;'
> Type=oneshot
> [Install]
> WantedBy=shutdown.target
> ```
> 
> Enable the "service" for `k3s-agent.service` (will also work for `k3s` on the master ):
> 
> ```
> sudo systemctl enable cgroup-kill-on-shutdown@k3s-agent.service.service
> 
> # or, on the master:  sudo systemctl enable cgroup-kill-on-shutdown@k3s.service.service
> ```
> 
> I've written a long winding explanation [here](https://raby.sh/debugging-90s-hangs-during-shutdown-on-ubuntu-2004.html) but in brief, what happens is that since `killmode=process` is used, all the container processes end up staying alive when k3s is brought down. Which is a good thing tm
> 
> However, during shutdown, systemd will signal all remaining processes and wait for `DefaultTimeoutStopSec` for them to die. This is **always** 90s during the last shutdown phase with systemd v245. It is a bug in systemd v245 shipped with ubuntu 20.04 and was [fixed in september 2020](https://github.com/systemd/systemd/commit/7d9eea2bd3d4f83668c7a78754d201b22)
> 
> What I used to do was to set `DefaultTimeoutStopSec=5s` in `/etc/systemd/system.conf` and it worked fine, but on ubuntu 20.04 it doesn't.
> 
> Since there's little chance this fix will make it back into 20.04, the above "service" will perform round of SIGTERM, wait 5s, then proceed with SIGKILL to finish k3s's process cleanup during shutdown. The sleep can be tweaked to suit your services need (something matching `terminationGracePeriod` perhaps)
> 
> Hope it helps.

This will only work with unified cgroups though as for example I don't have `/sys/fs/cgroup/unified/system.slice/` to begin with. :(

**Comment by damonmaria** [comment]
I could not get the systemd unit examples above to work (on Ubuntu 22.04). Instead the following achieved it for me:
```
[Unit]
Description=Kill containerd-shims on shutdown
Before=k3s.service

[Service]
Type=oneshot
RemainAfterExit=true
ExecStop=/usr/local/bin/k3s-killall.sh

[Install]
WantedBy=multi-user.target
```

`Before` needs to change to `k3s-agent.service` on agent nodes.

**Comment by brandond** [comment]
Yes, good catch. You will need to adapt the example for agent nodes. The server and agent use different service names.

**Comment by samip5** [comment]
> `Before` needs to change to `k3s-agent.service` on agent nodes.

Unless of course one uses k3s ansible role which names them both as k3s.service. :)

**Comment by kub3let** [comment]
> Here is a k3s version of [#2400 (comment)](https://github.com/k3s-io/k3s/issues/2400#issuecomment-1041165341):
> 
> ```
> [Unit]
> Description=Kill containerd-shims on shutdown
> DefaultDependencies=false
> Before=shutdown.target umount.target
> 
> [Service]
> ExecStart=/usr/local/bin/k3s-killall.sh
> Type=oneshot
> 
> [Install]
> WantedBy=shutdown.target
> ```
> 
> Put the file to `/etc/systemd/system/shutdown-k3s.service` and then enable the service using
> 
> ```
> systemctl enable shutdown-k3s.service
> ```
> 
> Also note that this service name `shutdown-k3s` shall not start with `k3s-`, otherwise the `k3s-killall.sh` script would try to stop it and cause problems.

---

Can confirm this also works if you get the message `A stop job is running for libcontainer...`

If you use the k3s ansible role you need to extract k3s-killall.sh from https://github.com/k3s-io/k3s/blob/d9f40d4f5b4776164322035499fabedea77f5f52/install.sh#L666-L743
